src/core/renderer.js

```diff
@@ -165,11 +165,35 @@ class Renderer {
 
         viewportExpressions.forEach(expr => expr._resetViewportAgg());
 
+        const viewportExpressionsLength = viewportExpressions.length;
+
         // Avoid acumulating the same feature multiple times keeping a set of processed features (same feature can belong to multiple dataframes).
         const processedFeaturesIDs = new Set();
 
+
+        this.aspect= this.gl.canvas.width / this.gl.canvas.height;
         dataframes.forEach(dataframe => {
+            const propertyNames = Object.keys(dataframe.properties);
+            const propertyNamesLength = propertyNames.length;
+
             for (let i = 0; i < dataframe.numFeatures; i++) {
+                let feature;
```

Comment:
Shouldn't all this be encapsulated in a Dataframe method?

Reply:
In fact this logic were in the `_featureFromDataFrame` method